diff --git a/ClearScriptV8/V8ContextImpl.cpp b/ClearScriptV8/V8ContextImpl.cpp
index 8ae90e37..e7d9c798 100644
--- a/ClearScriptV8/V8ContextImpl.cpp
+++ b/ClearScriptV8/V8ContextImpl.cpp
@@ -431,6 +431,8 @@ V8ContextImpl::V8ContextImpl(SharedPtr<V8IsolateImpl>&& spIsolateImpl, const Std
             throw FromMaybeFailure();
         }

+        m_hContext->SetSecurityToken(m_spIsolateImpl->GetContextSecurityToken());
+
         auto hContextImpl = CreateExternal(this);
         auto globalFlags = V8IsolateImpl::GetGlobalFlags();

diff --git a/ClearScriptV8/V8IsolateImpl.cpp b/ClearScriptV8/V8IsolateImpl.cpp
index 58e9fba8..a63056c8 100644
--- a/ClearScriptV8/V8IsolateImpl.cpp
+++ b/ClearScriptV8/V8IsolateImpl.cpp
@@ -520,6 +520,7 @@ V8IsolateImpl::V8IsolateImpl(const StdString& name, const v8::ResourceConstraint
             m_upIsolate->SetCaptureStackTraceForUncaughtExceptions(true, 64, v8::StackTrace::kDetailed);

             m_hHostObjectHolderKey = CreatePersistent(CreatePrivate());
+            m_hContextSecurityToken = CreatePersistent(v8::Symbol::New(m_upIsolate.get()));

             if (::HasFlag(options.Flags, Flags::EnableDebugging))
             {
@@ -1659,6 +1660,7 @@ V8IsolateImpl::~V8IsolateImpl()
     }

     Dispose(m_hHostObjectHolderKey);
+    Dispose(m_hContextSecurityToken);

     m_upIsolate->SetHostImportModuleDynamicallyCallback(static_cast<v8::HostImportModuleDynamicallyCallback>(nullptr));
     m_upIsolate->SetHostInitializeImportMetaObjectCallback(nullptr);
diff --git a/ClearScriptV8/V8IsolateImpl.h b/ClearScriptV8/V8IsolateImpl.h
index 175539a3..3ff22b08 100644
--- a/ClearScriptV8/V8IsolateImpl.h
+++ b/ClearScriptV8/V8IsolateImpl.h
@@ -151,6 +151,7 @@ public:

     const StdString& GetName() const { return m_Name; }
     const Persistent<v8::Private>& GetHostObjectHolderKey() const { return m_hHostObjectHolderKey; }
+    v8::Local<v8::Symbol> GetContextSecurityToken() const { return m_hContextSecurityToken.CreateLocal(m_upIsolate.get()); }
     const V8DocumentInfo* GetDocumentInfo() const { return m_pDocumentInfo; }

     v8::Local<v8::Context> CreateContext(v8::ExtensionConfiguration* pExtensionConfiguation = nullptr, v8::Local<v8::ObjectTemplate> hGlobalTemplate = v8::Local<v8::ObjectTemplate>(), v8::Local<v8::Value> hGlobalObject = v8::Local<v8::Value>())
@@ -669,6 +670,7 @@ private:
     UniqueDisposePtr<v8::Isolate> m_upIsolate;
     UniqueDisposePtr<v8::CpuProfiler> m_upCpuProfiler;
     Persistent<v8::Private> m_hHostObjectHolderKey;
+    Persistent<v8::Symbol> m_hContextSecurityToken;
     RecursiveMutex m_Mutex;
     std::list<ContextEntry> m_ContextEntries;
     SimpleMutex m_DataMutex;
